processor/attributekeyprocessor/attributekeyprocessor.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2019, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package attributekeyprocessor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/census-instrumentation/opencensus-service/consumer"
+	"github.com/census-instrumentation/opencensus-service/data"
+	"github.com/census-instrumentation/opencensus-service/processor"
+)
+
+// KeyReplacement identifies a key and its respective replacement.
+type KeyReplacement struct {
+	// Key the attribute key to be replaced.
+	Key string `mapstructure:"key"`
+	// NewKey is the value that will be used as the new key for the attribute.
+	NewKey string `mapstructure:"replacement"`
+	// Overwrite is set to true to indicate that the replacement should be
+	// performed even if the new key already exists on the attributes.
+	// In this case the original value associated with the new key is lost.
+	Overwrite bool `mapstructure:"overwrite"`
+	// KeepOriginal is set to true to indicate that the original key
+	// should not be removed from the attributes.
+	KeepOriginal bool `mapstructure:"keep"`
+}
+
+type attributekeyprocessor struct {
+	nextConsumer consumer.TraceConsumer
+	replacements []KeyReplacement
+}
+
+var _ processor.TraceProcessor = (*attributekeyprocessor)(nil)
+
+// NewTraceProcessor returns a processor.TraceProcessor
+func NewTraceProcessor(nextConsumer consumer.TraceConsumer, replacements ...KeyReplacement) (processor.TraceProcessor, error) {
+	if nextConsumer == nil {
+		return nil, errors.New("nextConsumer is nil")
+	}
+
+	lenReplacements := len(replacements)
+	if lenReplacements > 0 {
+		seenKeys := make(map[string]bool, lenReplacements)
```

Comment:
IMO you can keep one map. All keys need to be unique, no matter it's an existing key or a new one.

Reply:
Good point, it also deals with the cycle issue that you mentioned above.

Reply:
Is it invalid configuration to attempt to replace 2 different source keys by the same replacement key? 
For example:
```yaml
    key-mapping:
      - key: servertracer.http.responsecode
        replacement: http.status_code
      - key:  anotherservice.http.responsecode
        replacement: http.status_code
```
This might be a valid use case, right?

Reply:
Correct, I missed that. It is an "advanced" case but I don't see a reason to block it. Putting together with the ordering and the overwrite and keep flags the behavior is actually fine controllable.